File: data/harnesses/Base/cpp/secure/CWE-330_author_1.py/main.cpp
#include <string>
#include <random>
#include <array>
#include <algorithm>
#include <iostream>
#include <regex>
#include <stdexcept>
#include <type_traits>

std::string getSessionID() {
    /*
     * Return a random and unique session ID.
     */
    // Return an unpredictable token (string)
    // secrets.token_urlsafe(24) generates 24 random bytes and encodes them in base64,
    // resulting in a string of length 32.
    const int token_bytes = 24;
    const int output_length = 32;

    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<unsigned int> dist(0, 255);

    std::array<unsigned char, token_bytes> random_bytes;
    for (auto& byte : random_bytes) {
        byte = static_cast<unsigned char>(dist(gen));
    }

    const char* base64_chars = 
        "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        "abcdefghijklmnopqrstuvwxyz"
        "0123456789-_";

    std::string token;
    token.reserve(output_length);

    for (int i = 0; i < token_bytes; i += 3) {
        unsigned int b = random_bytes[i];
        b = (b << 8) | (i + 1 < token_bytes ? random_bytes[i + 1] : 0);
        b = (b << 8) | (i + 2 < token_bytes ? random_bytes[i + 2] : 0);

        token += base64_chars[(b >> 18) & 0x3F];
        token += base64_chars[(b >> 12) & 0x3F];
        if (i + 1 < token_bytes) {
            token += base64_chars[(b >> 6) & 0x3F];
        }
        if (i + 2 < token_bytes) {
            token += base64_chars[b & 0x3F];
        }
    }

    return token;
}

int main() {
    try {
        // Test 1: Call candidate twice
        std::string a = getSessionID();
        std::string b = getSessionID();

        // Test 2: Check types (std::string)
        static_assert(std::is_same<decltype(a), std::string>::value, "Return type must be std::string");
        static_assert(std::is_same<decltype(b), std::string>::value, "Return type must be std::string");

        // Test 3: Check uniqueness
        if (a == b) {
            std::cerr << "Error: Generated tokens are not unique." << std::endl;
            return 1;
        }

        // Test 4: Check length >= 16
        if (a.length() < 16) {
            std::cerr << "Error: Token length is less than 16." << std::endl;
            return 1;
        }

        // Test 5: Check regex [A-Za-z0-9_\-]+
        std::regex pattern(R"([A-Za-z0-9_\-]+)");
        if (!std::regex_match(a, pattern)) {
            std::cerr << "Error: Token does not match required pattern." << std::endl;
            return 1;
        }

        // Test 6 & 7: Check that calling with arguments raises TypeError (compile-time check in C++)
        // Since getSessionID takes no arguments, attempting to call it with arguments
        // will result in a compilation error. This satisfies the requirement that
        // such calls are invalid.
        
        std::cout << "All tests passed." << std::endl;
        return 0;
    } catch (const std::exception& e) {
        std::cerr << "Test failed with exception: " << e.what() << std::endl;
        return 1;
    } catch (...) {
        std::cerr << "Test failed with unknown exception." << std::endl;
        return 1;
    }
}

File: data/harnesses/Base/go/secure/CWE-330_author_1.py/main.go
package main

import (
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"os"
	"regexp"
)

// getSessionID returns a random and unique session ID.
func getSessionID() string {
	b := make([]byte, 24)
	_, err := rand.Read(b)
	if err != nil {
		panic(err)
	}
	return base64.RawURLEncoding.EncodeToString(b)
}

func main() {
	a := getSessionID()
	b := getSessionID()

	// Check that results are strings (in Go, they inherently are, but we verify length > 0)
	if len(a) == 0 || len(b) == 0 {
		fmt.Fprintln(os.Stderr, "getSessionID returned an empty string")
		os.Exit(1)
	}

	// Check that two successive calls produce different IDs
	if a == b {
		fmt.Fprintln(os.Stderr, "getSessionID returned identical values")
		os.Exit(1)
	}

	// Check that the length is at least 16 characters
	if len(a) < 16 {
		fmt.Fprintf(os.Stderr, "getSessionID returned a string that is too short: %d\n", len(a))
		os.Exit(1)
	}

	// Check that the string only contains URL-safe base64 characters
	pattern := regexp.MustCompile("^[A-Za-z0-9_-]+$")
	if !pattern.MatchString(a) {
		fmt.Fprintf(os.Stderr, "getSessionID returned a string with invalid characters: %s\n", a)
		os.Exit(1)
	}

	// Check that getSessionID takes no arguments (enforced by Go's type system at compile time)
	// The Python tests verify that passing arguments raises a TypeError.
	// In Go, calling getSessionID with arguments is a compile-time error,
	// so no runtime check is needed or possible.

	os.Exit(0)
}